src/freenet/support/MultiValueTable.java

```diff
@@ -1,230 +1,260 @@
+     * @param initialSize table initial size
+     * @param unused this parameter was used as initial value collection size, and it is not unused anymore.
+     *               Values collections grow according to Java collection rules.
+     */
+    @Deprecated
+    public MultiValueTable(int initialSize, int unused) {
+        this(initialSize);
     }
 
     public static <K, V> MultiValueTable<K, V> from (K[] keys, V[] values) {
-      if (keys.length != values.length) {
-        throw new IllegalArgumentException(String.format(
-            "keys and values must contain the same number of values, but there are %d keys and %d values",
-            keys.length,
-            values.length));
-      }
-      MultiValueTable<K, V> table = new MultiValueTable<>();
-      for (int i = 0; i < keys.length; i++) {
-        table.put(keys[i], values[i]);
-      }
-      return table;
+        if (keys.length != values.length) {
+            throw new IllegalArgumentException(String.format(
+                "keys and values must contain the same number of values, but there are %d keys and %d values",
+                keys.length,
+                values.length));
+        }
+        MultiValueTable<K, V> table = new MultiValueTable<>(keys.length);
+        for (int i = 0; i < keys.length; i++) {
+            table.put(keys[i], values[i]);
+        }
+        return table;
+    }
+
+    @SafeVarargs
+    public static <K, V> MultiValueTable<K, V> from(K key, V... values) {
+        return from(key, Arrays.asList(values));
+    }
+
+    public static <K, V> MultiValueTable<K, V> from(K key, Collection<V> values) {
+        MultiValueTable<K, V> table = new MultiValueTable<>(1);
+        table.putAll(key, values);
+        return table;
     }
 
     public void put(K key, V value) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            if (v == null) {
-                v = new Vector<V>(ies);
-                table.put(key, v);
+        this.table.compute(key, (k, list) -> {
+            if (list == null) {
+                list = new CopyOnWriteArrayList<>();
             }
-            v.addElement(value);
-        }
+            list.add(value);
+            return list;
+        });
+    }
+
+    public void putAll(K key, Collection<V> elements) {
+        this.table.compute(key, (k, list) -> {
+            if (list == null) {
+                list = new CopyOnWriteArrayList<>(elements);
+            } else {
+                list.addAll(elements);
+            }
+            return list;
+        });
     }
 
     /**
      * Returns the first element for this key.
      */
     public V get(K key) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            return (v == null ?
-                    null :
-                    v.firstElement());
+        return getFirst(key);
+    }
+
+    /**
+     * Returns the first element for this key.
+     */
+    public V getFirst(K key) {
```

Comment:
Now this is just a duplicate of `get`. Is there any real need for extending the public API with this method? If so, please consider deprecating the method it replaces.

Reply:
New method looks more explicit for me.
Besides that there is no that big need and old method may be still used.
What variant do you prefer more?

Reply:
`getFirst` make way more sense indeed (not that any of the other methods of this class make much sense, but that's all hindsight)

We should not have 2 non-deprecated methods doing the exact same thing, because then things get confusing for developers - which one should you use? I'm all for `getFirst`, but then let's deprecate `get` and refer to `getFirst` in the `@deprecated` notice.

Reply:
Added more deprecation comments there.